### PR TITLE
refactor: Reorganize handling of requests and file parsing

### PR DESCRIPTION
**Pull request by slavek-kucera**

### CODE REVIEW

File: parser_library/src/workspace_manager_impl.h [hunk 1]
```diff
@@ -203,48 +392,73 @@ class workspace_manager::impl final : public diagnosable_impl
...
-        if (cancel_ && *cancel_)
-        {
-            definition_result = { pos, doc_loc };
-            r.provide(position_uri(definition_result));
-            return;
-        }
+        int code;
+        const char* msg;
+    } static constexpr request_cancelled { -32800, "Canceled" }, request_failed { -32803, "Unknown reason" },
+        removing_workspace { -32803, "Workspace removal in progress" };
```
Comment (jirimosinger) [inline parser_library/src/workspace_manager_impl.h:400]: Is it intentional to have -32803 for two different messages?

File: utils/include/utils/scope_exit.h [hunk 2]
```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright (c) 2021 Broadcom.
```
[the remark this answers is not in the document]
Comment (slavek-kucera) [reply utils/include/utils/scope_exit.h:2]: This file was just moved and updated, it is not new.

File: language_server/src/main.cpp [hunk 3]
```diff
@@ -53,41 +51,64 @@ class main_program : public json_sink
     dap::session_manager dap_sessions;
     virtual_file_provider virtual_files;
 
+    void reply(const nlohmann::json& message) final { json_output.write(message); }
```
[the remark this answers is not in the document]
Comment (slavek-kucera) [reply language_server/src/main.cpp:54]: we could, however there would be benefit only if a transformation is applied to the json. In this case, I think it is just turned into a string and written out.

File: parser_library/src/workspace_manager_impl.h [hunk 4]
```diff
@@ -99,15 +103,31 @@ class workspace_manager::impl final : public diagnosable_impl
...
+    {
+        workspaces::workspace* ws;
+
+        std::function<void(bool)> action; // true on workspace removal
+        std::function<bool()> validator; // maybe empty
+
+        work_item_type request_type;
+        bool workspace_removed = false;
+
+        bool is_valid() const { return !validator || validator(); }
```
[the remark this answers is not in the document]
Comment (slavek-kucera) [reply parser_library/src/workspace_manager_impl.h:122]: The idea was that the item is assumed always valid if not validation routine is provided, so the code should be fine.

File: parser_library/src/workspace_manager_impl.h (hunk 1 above)
Comment (slavek-kucera) [reply parser_library/src/workspace_manager_impl.h:400]: the -32803 means "request failed", the message is just a reason. I included the `request_failed` just to indicate where it came from.

File: language_server/src/server.h [hunk 5]
```diff
@@ -15,8 +15,11 @@
 #ifndef HLASMPLUGIN_LANGUAGESERVER_SERVER
 #define HLASMPLUGIN_LANGUAGESERVER_SERVER
 
+#include <chrono>
```
[the remark this answers is not in the document]
Comment (slavek-kucera) [reply language_server/src/server.h:18]: regarding the json header, it is already required for the response provider (in the fwd form), so it should be fine.

File: language_server/src/main.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (jirimosinger) [reply language_server/src/main.cpp:54]: I've got a feeling that we're making copies of `nlohmann::json` somewhere along the call chain in our code. But we can take a closer look next time

**Review by jirimosinger — APPROVED** [review]

File: language_server/src/main.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (slavek-kucera) [reply language_server/src/main.cpp:54]: At this point the full header should be only included in cpp files. The only exception seems to be the benchmark subproject - we can fix that.

**Pull request merged**